dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/report/ValidationErrorReporter.java

```diff
@@ -0,0 +1,149 @@
+ */
+
+import lombok.Data;
+import org.hisp.dhis.tracker.ValidationMode;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.domain.TrackerDto;
+import org.hisp.dhis.tracker.validation.TrackerImportValidationContext;
+import org.hisp.dhis.tracker.validation.ValidationFailFastException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+@Data
+public class ValidationErrorReporter
+{
+    private final List<TrackerErrorReport> reportList;
+
+    private final boolean isFailFast;
+
+    private final TrackerImportValidationContext validationContext;
+
+    private final Class<?> mainKlass;
+
+    private final AtomicInteger listIndex = new AtomicInteger( 0 );
+
+    private String mainId;
+
+    public ValidationErrorReporter( TrackerImportValidationContext context, Class<?> mainKlass )
+    {
+        this.validationContext = context;
+        this.reportList = new ArrayList<>();
+        this.isFailFast = validationContext.getBundle().getValidationMode() == ValidationMode.FAIL_FAST;
+        this.mainKlass = mainKlass;
+    }
+
+    private ValidationErrorReporter( TrackerImportValidationContext context, Class<?> mainKlass, boolean isFailFast,
+        int listIndex )
+    {
+        this.validationContext = context;
+        this.reportList = new ArrayList<>();
+        this.isFailFast = isFailFast;
+        this.mainKlass = mainKlass;
+        this.listIndex.set( listIndex );
+    }
+
+    public boolean hasErrors()
+    {
+        return !this.reportList.isEmpty();
+    }
+
+    public static TrackerErrorReport.TrackerErrorReportBuilder newReport( TrackerErrorCode errorCode )
+    {
+        return TrackerErrorReport.builder().errorCode( errorCode );
+    }
+
+    public void addError( TrackerErrorReport.TrackerErrorReportBuilder builder )
+    {
+        builder.mainKlass( this.mainKlass );
+        builder.listIndex( this.listIndex.get() );
+
+        if ( this.mainId != null )
+        {
+            builder.mainId( this.mainId );
+        }
+
+        getReportList().add( builder.build( this.validationContext.getBundle() ) );
+
+        if ( isFailFast() )
+        {
+            throw new ValidationFailFastException( getReportList() );
+        }
+    }
+
+    public <T extends TrackerDto> ValidationErrorReporter fork()
+    {
+        return fork( null );
+    }
+
+    public <T extends TrackerDto> ValidationErrorReporter fork( T dto )
+    {
+        ValidationErrorReporter fork = new ValidationErrorReporter( this.validationContext, this.mainKlass,
+            isFailFast(),
+            listIndex.incrementAndGet() );
+
+        if ( dto == null )
+        {
+            fork.mainId = this.mainId;
+            return fork;
+        }
+
+        // TODO: Use interface method to build name?
```

Comment:
Yes, TrackerDto should have something like a getId method so we could remove the instanceof here